xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -77,27 +132,112 @@ public Server start() throws IOException {
       throw new RuntimeException(ex);
     }
     delegate.start();
+    currentServingState = ServingState.STARTED;
     xdsServingStatusListener.onServing();
     return this;
   }
 
-  private Future<EnvoyServerProtoData.DownstreamTlsContext> addServerWatcher() {
-    final SettableFuture<EnvoyServerProtoData.DownstreamTlsContext> settableFuture =
-        SettableFuture.create();
+  @VisibleForTesting public ServingState getCurrentServingState() {
+    return currentServingState;
+  }
+
+  private SettableFuture<Throwable> addServerWatcher() {
+    final SettableFuture<Throwable> future = SettableFuture.create();
     serverWatcher =
         new XdsClientWrapperForServerSds.ServerWatcher() {
           @Override
-          public void onError(Throwable throwable) {
-            xdsServingStatusListener.onNotServing(throwable);
+          public void onError(Throwable throwable, boolean isAbsent) {
+            if (currentServingState == ServingState.STARTING) {
```

Comment:
There could be a race between checking the value and currentServingState getting updated from STARTING to STARTED. If the checking the value here wins the race, it may change currentServingState to SHUTDOWN, and then changed to STARTED by another thread.

Reply:
(also addressing your comment for line 162):

Yes, good point. There are 3 possible sources (in terms of threads) of updates to `currentServingState`:
- user thread for the updates happening in `start()` or `shutdown*()` 
- the `XdsClient` thread sending listener updates
- the Timer task that's periodically trying to do `delegate.start()` because of `RetriableErrorInDelegateStart`

So the race conditions that need to be resolved are the ones between the XdsClient thread and the Timer task and the `shutdown*()` call and one of XdsClient or Timer task. Let me try to address this.

Reply:
synchronizing on `ServerWrapperForXds.this` now. Also added a couple of other things like `AtomicReference` for `ScheduledFuture<?> future` inside the `StartRetryTask` and more synchronization in `cleanupStartRetryTaskAndShutdownDelegateAndXdsClient`